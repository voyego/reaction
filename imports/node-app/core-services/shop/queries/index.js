/* eslint-disable node/no-deprecated-api */
/* TODO: revisit `url.parse` throughout Reaction */

export default {
  shopById: (context, _id) => context.dataLoaders.Shops.load(_id),
  shopBySlug: (context, slug) => context.collections.Shops.findOne({ slug }),
  primaryShop: async (context) => {
    const { collections } = context;
    const { Shops } = collections;
    const shop = await Shops.findOne({ shopType: "primary" });
    return shop;
  }
};
