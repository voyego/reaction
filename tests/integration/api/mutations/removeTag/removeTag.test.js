import Factory from "/imports/test-utils/helpers/factory";
import TestApp from "/imports/test-utils/helpers/TestApp";
import { encodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";
import { encodeTagOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/tag";
import RemoveTagMutation from "./RemoveTagMutation.graphql";
import Logger from "@reactioncommerce/logger";

let testApp;
let shopId;
let removeTag;
let mockTagsAccount;
let tagInput;
let fakeTag;

const accountInternalId = "123";

beforeAll(async () => {
  testApp = new TestApp();
  await testApp.start();
  shopId = await testApp.insertPrimaryShop();

  mockTagsAccount = Factory.Account.makeOne({
    roles: {
      [encodeShopOpaqueId(shopId)]: ["owner"]
    }
  });
  await testApp.createUserAndAccount(mockTagsAccount);

  removeTag = testApp.mutate(RemoveTagMutation);

  fakeTag = Factory.Tag.makeOne({
    _id: encodeTagOpaqueId(accountInternalId),
    shopId: encodeShopOpaqueId(shopId)
  });

  tagInput = {
    id: encodeTagOpaqueId(fakeTag._id),
    shopId: encodeShopOpaqueId(fakeTag.shopId)
  };
});

afterAll(async () => {
  await testApp.collections.Shops.deleteMany({});
  await testApp.stop();
});
beforeEach(async () => {
  await testApp.collections.Tags.insertOne(fakeTag);
});
afterEach(async () => {
  await testApp.collections.Tags.deleteMany({});
});

describe("unauthorized user", () => {
  let logLevel;
  beforeAll(async () => {
    await testApp.setLoggedInUser({ _id: accountInternalId });
    logLevel = Logger.level();
    Logger.level("FATAL");
  });
  afterAll(async () => {
    await testApp.clearLoggedInUser();
    Logger.level(logLevel);
  });
  test("cannot remove tag", async () => {
    expect.assertions(1);
    try {
      await removeTag(tagInput);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe("authorized user", () => {
  beforeAll(async () => {
    await testApp.setLoggedInUser(mockTagsAccount);
  });
  afterAll(async () => {
    await testApp.clearLoggedInUser();
  });
  test("can remove tag", async () => {
    try {
      let removedTag = await testApp.collections.Tags.findOne({ _id: fakeTag._id, shopId: fakeTag.shopId });
      // tag is returned unencoded diretly from mongoDB but is returned encoded from the mutation
      removedTag._id = encodeTagOpaqueId(removedTag._id);
      removedTag.heroMediaUrl = `https://shop.fake.site/${removedTag.heroMediaUrl}`;
      expect(removedTag).not.toBeNull();
      const result = await removeTag(tagInput);
      // mongoDB tag returns more fields than removeTag mutation, so objectContaining is to ensure returned fields match from db entry.
      expect(removedTag).toEqual(expect.objectContaining(result.removeTag.tag));
      removedTag = await testApp.collections.Tags.findOne({ _id: fakeTag._id, shopId: fakeTag.shopId });
      expect(removedTag).toBeNull();
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });
});
