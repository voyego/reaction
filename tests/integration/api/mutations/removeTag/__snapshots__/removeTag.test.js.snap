// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unauthorized user cannot remove tag 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "FORBIDDEN",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "User does not have permission",
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "User does not have permission",
    "path": Array [
      "removeTag",
    ],
  },
]
`;
